portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -19,53 +19,73 @@
 import com.google.appengine.api.datastore.Entity;
 import com.google.appengine.api.datastore.PreparedQuery;
 import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
 import com.google.gson.Gson;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 import java.util.ArrayList;
+import java.util.List;
+import java.util.Collections;
 
 
 /**
-  * Servlet that returns some example content. 
+  * Servlet that returns comments in database. 
   * Note: We first added the Gson library dependency to pom.xml.
   */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
-
+  int maxNumComments = 1;
+  ArrayList<String> msglist = new ArrayList<String>();
   String comment = "Comment";
+  String contentProperty = "content";
+  String timestampProperty = "timestamp";
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String inputNumComment = "numComment";
+    maxNumComments = Integer.parseInt(request.getParameter(inputNumComment));
+    
     // Create the query and prepared query to load comment entities from database
-    Query query = new Query(comment);
+    Query query = new Query(comment).addSort(timestampProperty, SortDirection.DESCENDING);
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
     
-    // Add all comment contents to the msgList  
+    // Add all comment contents to the msglist  
     ArrayList<String> msglist = new ArrayList<String>();
     for (Entity comment:results.asIterable()) {
-      String commentMsg = (String)comment.getProperty("content");
+      String commentMsg = (String)comment.getProperty(contentProperty);
       msglist.add(commentMsg);
     }
+    Collections.shuffle(msglist);
+
+    // Limit number of comments;
+    List<String> updatedStrList;
+    if (msglist.size() > maxNumComments) {
+        updatedStrList = msglist.subList(0,maxNumComments);
+    } else {
+        updatedStrList = msglist;
+    }
     response.setContentType("application/json;");
-    response.getWriter().println(new Gson().toJson(msglist));
+    response.getWriter().println(new Gson().toJson(updatedStrList)) ;
   }
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Get the input from the form.
-    String inputMsg = request.getParameter("comment-input");
+    String inputParams = "comment-input";
+    String inputMsg = request.getParameter(inputParams);
     if (!inputMsg.isEmpty()) {
```

Comment:
I don't think isEmpty() works for a null argument. ie. if getParamater returns a null

Reply:
The official docs for HttpServeletRequest.getParameter() says: Returns the value of a reuqest as a String, "or null if the parameter does not exists". The case I want to handle here is when the input comment message is empty ( like the user does not put any comment message but click on the submit button). This will return non-null string but just a "". I have a discussion about this with Cis at step 5 and we reach to the conclusion that .isEmpty() is fine for now. Maybe any further recommentdations?

Reply:
We should check that parameters are null here. We cannot assume that this Post will always set the parameters correctly. Hypothetically, you could change the front end and we could call this with a null parameter